File: android_native/app/src/main/java/com/smartexpense/manager/domain/repository/ExpenseRepository.kt
package com.smartexpense.manager.domain.repository

import com.smartexpense.manager.domain.model.Expense
import com.smartexpense.manager.domain.model.Category
import java.util.Date

interface ExpenseRepository {
    suspend fun getExpenses(): Result<List<Expense>>
    suspend fun getExpenseById(id: String): Result<Expense?>
    suspend fun saveExpense(expense: Expense): Result<Unit>
    suspend fun deleteExpense(expense: Expense): Result<Unit>
    suspend fun getExpensesByDate(date: Date): Result<List<Expense>>
    suspend fun getExpensesByCategory(category: Category): Result<List<Expense>>
    suspend fun getTotalAmount(): Result<Double>
}

File: flutter/lib/domain/repositories/expense_repository.dart
import 'package:smart_expense_manager/core/result.dart';
import 'package:smart_expense_manager/domain/models/expense.dart';

abstract class ExpenseRepository {
  Future<Result<List<Expense>>> getExpenses();
  Future<Result<Expense?>> getExpenseById(String id);
  Future<Result<void>> saveExpense(Expense expense);
  Future<Result<void>> deleteExpense(Expense expense);
  Future<Result<List<Expense>>> getExpensesByDate(DateTime date);
  Future<Result<List<Expense>>> getExpensesByCategory(Category category);
  Future<Result<double>> getTotalAmount();
}
